C error when handling optionals during variable assignment to `shared`

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:**  V 0.2.2 043f642.bcb3992
**OS:** windows, Microsoft Windows 10 Pro v19042 64-Bit

**What did you do?**
```v
module main

struct ABC {
    s string = ''
}

fn test_shared() {
    shared abc := foo() or { panic('scared') }
}

fn foo() ?ABC {
    a := ABC{}
    a.bar() ?
    return a
}

fn (a ABC) bar() ? {
    println(a.s)
    panic('boo!')
}
```

**What did you expect to see?**
The `or` block handles the `error`/`panic` propagated by `bar()` and the code exits due to the `panic`.
 
**What did you see instead?**
The following compiler error + C code; see near `struct __shared__void`
```
C:/Users/XENIAL~1/AppData/Local/Temp/v/shared_crash_test.3078357629458636315.tmp.c:426: warning: WINVER redefined
In file included from C:/Users/XENIAL~1/AppData/Local/Temp/v/shared_crash_test.3078357629458636315.tmp.c:864:
C:/users/xenialdan/desktop/v/v/thirdparty/stdatomic/win/atomic.h:123: warning: InterlockedIncrement64 redefined
In file included from C:/Users/XENIAL~1/AppData/Local/Temp/v/shared_crash_test.3078357629458636315.tmp.c:878:
c:/users/xenialdan/desktop/v/v/thirdparty/tcc/include/winapi/synchapi.h:13: warning: CreateEvent redefined
In file included from C:/Users/XENIAL~1/AppData/Local/Temp/v/shared_crash_test.3078357629458636315.tmp.c:878:
c:/users/xenialdan/desktop/v/v/thirdparty/tcc/include/winapi/synchapi.h:117: warning: OpenEvent redefined
In file included from C:/Users/XENIAL~1/AppData/Local/Temp/v/shared_crash_test.3078357629458636315.tmp.c:878:
c:/users/xenialdan/desktop/v/v/thirdparty/tcc/include/winapi/synchapi.h:167: warning: CreateMutex redefined
C:/Users/XENIAL~1/AppData/Local/Temp/v/shared_crash_test.3078357629458636315.tmp.c:2197: error: invalid type for 'val'
builder error:
==================
C error. This should never happen.
```

```c
typedef struct thread_arg_main__read_input {
	__shared__bstone__Settings* arg1;
} thread_arg_main__read_input;
u32 main__read_input_thread_wrapper(thread_arg_main__read_input *arg);

// V shared types:
struct __shared__bstone__Settings { bstone__Settings val; sync__RwMutex mtx; };
struct __shared__main__ABC { main__ABC val; sync__RwMutex mtx; };
struct __shared__void { void val; sync__RwMutex mtx; };

// V Option_xxx definitions:
struct Option_int {
	byte state;
	IError err;
	byte data[sizeof(int)];
};
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2903,7 +2903,7 @@ fn (mut g Gen) expr(node_ ast.Expr) {
 				node.return_type.clear_flag(.optional)
 			}
 			mut shared_styp := ''
-			if g.is_shared && !ret_type.has_flag(.shared_f) {
+			if g.is_shared && !ret_type.has_flag(.shared_f) && !g.inside_or_block {
 				ret_sym := g.table.sym(ret_type)
 				shared_typ := ret_type.set_flag(.shared_f)
 				shared_styp = g.typ(shared_typ)
@@ -2932,7 +2932,7 @@ fn (mut g Gen) expr(node_ ast.Expr) {
 				g.strs_to_free0 = []
 				// println('pos=$node.pos.pos')
 			}
-			if g.is_shared && !ret_type.has_flag(.shared_f) {
+			if g.is_shared && !ret_type.has_flag(.shared_f) && !g.inside_or_block {
 				g.writeln('}, sizeof($shared_styp))')
 			}
 			// if g.autofree && node.autofree_pregen != '' { // g.strs_to_free0.len != 0 {

--- a/vlib/v/tests/shared_optional_test.v
+++ b/vlib/v/tests/shared_optional_test.v
@@ -1,0 +1,25 @@
+module main
+
+struct ABC {
+mut:
+	s string
+}
+
+fn test_shared_optional() {
+	shared abc := foo() or { panic('scared') }
+	rlock abc {
+		println(abc)
+		assert abc.s == 'hello'
+	}
+}
+
+fn foo() ?ABC {
+	mut a := ABC{}
+	a.bar()?
+	return a
+}
+
+fn (mut a ABC) bar() ? {
+	a.s = 'hello'
+	println(a.s)
+}

